### ISSUE TITLE
[Feature Request] Is it possible to align the right hand side of the mode-line precisely with a variable-pitch font?

### ISSUE THREAD

**Issue by Dima-369** [seagle0128/doom-modeline#494]
I know this is a rather minor thing as noted in the quote below, but I wondered if it is still possible because I try to use a variable-pitch font where possible.

Here was a previous patch/comment related to it in https://github.com/seagle0128/doom-modeline/issues/301#issuecomment-584762592

> I committed a patch to support different fonts in mode line. Just need to set mode-line and mode-line-inactive face. All icons should be Okay now. The right position of the segment may not be accurate but I think it's a minor issue.

Depending on which characters one uses in the right hand side, the offset to the right window margin is varying a lot:

![image](https://user-images.githubusercontent.com/15002298/146264402-81fe71da-88ed-40d6-97e3-cd1813714b02.png)

I tested with this code:

```emacs-lisp
(doom-modeline-mode 1)
(set-face-attribute 'mode-line nil
                    :font "Avenir" ; present on macOS but I suppose any other variable-pitch font would reproduce
                    :height 201)
```

---

I don't know if this is possible because `(string-width)` is used in `(doom-modeline-def-modeline`) which does not return pixels. And I suspect that it might be a slight performance hit as one needs to calculate the right hand side pixel width and then subtract it from the total window pixel width.

Still I am wondering if it could be implemented without impacting performance too much as all the 'big' non-Emacs code editors have it in their mode-line variant :)

---

This is possibly related to https://github.com/seagle0128/doom-modeline/issues/487

**Comment by seagle0128** [comment]
>I don't know if this is possible because (string-width) is used in (doom-modeline-def-modeline) which does not return pixels. And I suspect that it might be a slight performance hit as one needs to calculate the right hand side pixel width and then subtract it from the total window pixel width.

I don't think it's possible so far, unless new elisp functions are provided by Emacs. `string-width` is not able to calculate the width of variable-picth fonts. If you happen to know, please also let me know. Thanks!

**Comment by Gira-X** [comment]
There is `(shr-string-pixel-width)` which returns the real pixels the string takes up:

```emacs-lisp
(defface face-test
  `((t :height 300 ))
  "Test with larger height.")

(let ((shr-use-fonts t))
  (shr-string-pixel-width "test")
  ;; => 34
  (shr-string-pixel-width (propertize "test" 'face 'face-test))
  ;; => 50
  )
```

The source of the function is:

```emacs-lisp
(defun shr-string-pixel-width (string)
  (if (not shr-use-fonts)
      (length string)
    ;; Save and restore point across with-temp-buffer, since
    ;; shr-pixel-column uses save-window-excursion, which can reset
    ;; point to 1.
    (let ((pt (point)))
      (prog1
	  (with-temp-buffer
	    (insert string)
	    (shr-pixel-column))
	(goto-char pt)))))
```

Could this be used together with `(window-pixel-width)` in the mode-line?

**Comment by seagle0128** [comment]
What's the function ` shr-pixel-column`？

**Comment by Gira-X** [comment]
From `shr.el.gz`:

```emacs-lisp
(defun shr-pixel-column ()
  (if (not shr-use-fonts)
      (current-column)
    (if (not (get-buffer-window (current-buffer)))
	(save-window-excursion
          ;; Avoid errors if the selected window is a dedicated one,
          ;; and they just want to insert a document into it.
          (set-window-dedicated-p nil nil)
	  (set-window-buffer nil (current-buffer))
	  (car (window-text-pixel-size nil (line-beginning-position) (point))))
      (car (window-text-pixel-size nil (line-beginning-position) (point))))))
```

**Comment by seagle0128** [comment]
Thanks! I didn't realized it was built-in shr. I tried `shr-string-pixel-width` and it seems not working well with bold fonts. Can you check as well?

e.g. the both below are same width
```elisp
(doom-modeline-string-pixel-width (propertize "test"))
(doom-modeline-string-pixel-width (propertize "test" 'face 'bold))
```

**Comment by Gira-X** [comment]
It does work when using `defface` for me. Maybe your font has the same character width when bold? But `'face 'bold` should also work; I just have the bold weight removed from my `bold` face because I do not like the bold fonts in Emacs.

```emacs-lisp
(defface face-test1
  `((t :height 300
       :font "Avenir"
       :weight bold))
  "")

(defface face-test2
  `((t :height 300
       :font "Avenir"))
  "")

(message "%s - %s"
           (transient--pixel-width (propertize "test" 'face 'face-test1))
           (transient--pixel-width (propertize "test" 'face 'face-test2)))
;; => prints "54 - 50"
```

---

And I noticed how this function from `transient` (https://github.com/magit/transient) is a lot simpler but returns the same:

```emacs-lisp
(defun transient--pixel-width (string)
  (save-window-excursion
    (with-temp-buffer
      (insert string)
      (set-window-dedicated-p nil nil)
      (set-window-buffer nil (current-buffer))
      (car (window-text-pixel-size
            nil (line-beginning-position) (point))))))
```

**Comment by Gira-X** [comment]
```emacs-lisp
(transient--pixel-width (propertize "test"))
(transient--pixel-width (propertize "test" 'face 'bold))
```

I tried the code above with the default `bold` face and your code above returns differing string widths correctly, so it certainly works for `bold` faces!

**Comment by seagle0128** [comment]
Unfortunately it doesn't work on macOS. Both return the same result 😢  Thus will make the mode-lines different in active and inactive windows.

**Comment by Gira-X** [comment]
Hmm, I am on macOS 10.14.6 on Emacs on `GNU Emacs 27.2 (build 1, x86_64-apple-darwin18.5.0, Carbon Version 158 AppKit 1671.4) of 2021-04-05` and it works for me, regardless of the face or bold variants. Strange that it does not work for you :(

**Comment by Gira-X** [comment]
Can you run this and check if you see that the text is really visually wider in the bold variant?

```emacs-lisp
(defun test ()
  (interactive)
  (let ((shr-use-fonts t)
        (test-string "test"))
    (message "%s %d\n%s %d"
             test-string (shr-string-pixel-width test-string)
             (propertize test-string 'face 'bold)
             (shr-string-pixel-width (propertize test-string 'face 'bold)))))
```

![image](https://user-images.githubusercontent.com/15002298/146587443-baf0bbdf-16f9-4b55-bf2f-bbd7fcab54d5.png)

**Comment by seagle0128** [comment]
I got this on macOS

![image](https://user-images.githubusercontent.com/140797/146669239-44243021-256f-4b26-b6fb-f0b4905dabaa.png)

**Comment by Gira-X** [comment]
Your image above shows a monospace font.

Can you set a variable-pitch font as the default font of Emacs and then try again?

I just find it rather peculiar that it does not work for you!

```emacs-lisp
(set-face-attribute 'default nil :font "Helvetica")
```

**Comment by seagle0128** [comment]
It does work for variable-pitch fonts, while not for monospace fonts. The calculation of character width is still not accurate 😞

**Comment by deifactor** [comment]
I was poking a little more at this, and while I wasn't able to get an accurate width calculation, I did figure out that `beacon-mode` interferes with this; sometimes it causes the rhs to move all the way to the left (I assume because it's causing the computed width to be extremely large) and it causes the beacon itself to break.

**Comment by seagle0128** [comment]
I think it's working now. Closed by 1d400aa.

**Comment by erusyd** [comment]
> I think it's working now. Closed by [1d400aa](https://github.com/seagle0128/doom-modeline/commit/1d400aa88372962be6dad80182c244df3fe6e00b).

Emacs complains:
Error during redisplay: (eval (doom-modeline-format--main)) signaled (void-function string-pixel-width)

**Comment by seagle0128** [comment]
@erusyd  Sorry for the regression issue! It has been fixed in https://github.com/seagle0128/doom-modeline/commit/2ff72d794d6a7ed4c5c5ac8b0b314704424150bd.
